File: test-suite-server-generator/src/main/java/io/micronaut/openapi/test/api/RequestBodyController.java
package io.micronaut.openapi.test.api;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.util.List;

import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Post;
import io.micronaut.http.multipart.CompletedFileUpload;
import io.micronaut.openapi.test.model.*;

import jakarta.validation.Valid;
import reactor.core.publisher.Mono;

@Controller
public class RequestBodyController implements RequestBodyApi {

    @Post(value = "/echo", consumes = "text/plain", produces = "text/plain")
    public String echo(String request) {
        return request;
    }

    @Override
    public Mono<SimpleModel> sendSimpleModel(SimpleModel simpleModel) {
        return Mono.just(simpleModel);
    }

    @Override
    public Mono<Void> sendValidatedCollection(List<List<String>> requestBody) {
        return null;
    }

    @Override
    public Mono<List<SimpleModel>> sendListOfSimpleModels(List<SimpleModel> simpleModels) {
        return Mono.just(simpleModels);
    }

    @Override
    public Mono<ModelWithRequiredProperties> sendModelWithRequiredProperties(ModelWithRequiredProperties model) {
        return Mono.just(model);
    }

    @Override
    public Mono<DateModel> sendDateModel(DateModel model) {
        return Mono.just(model);
    }

    @Override
    public Mono<ColorEnum> sendEnum(String color) {
        return Mono.just(ColorEnum.fromValue(color.replace("\"", "")));
    }

    @Override
    public Mono<List<ColorEnum>> sendEnumList(
        List<@Valid ColorEnum> availableColors) {
        return Mono.just(availableColors);
    }

    @Override
    public Mono<ModelWithMapProperty> sendModelWithMapProperty(ModelWithMapProperty model) {
        return Mono.just(model);
    }

    @Override
    public Mono<Void> sendModelWithValidatedListProperty(ModelWithValidatedListProperty model) {
        return Mono.empty();
    }

    @Override
    public Mono<NestedModel> sendNestedModel(NestedModel model) {
        return Mono.just(model);
    }

    @Override
    public Mono<ModelWithInnerEnum> sendModelWithInnerEnum(ModelWithInnerEnum model) {
        return Mono.just(model);
    }

    @Override
    public Mono<Animal> sendModelWithDiscriminator(Animal model) {
        return Mono.just(model);
    }

    @Override
    public Mono<byte[]> sendBytes(byte[] bytes) {
        return Mono.just(bytes);
    }

    @Override
    public Mono<ModelWithEnumList> sendModelWithEnumList(ModelWithEnumList model) {
        return Mono.just(model);
    }

    @Override
    public Mono<byte[]> sendFile(CompletedFileUpload file) {
        return Mono.fromCallable(() -> {
            InputStream inputStream = file.getInputStream();
            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
            outputStream.write("name: ".getBytes());
            outputStream.write(file.getFilename().getBytes());
            outputStream.write(", content: ".getBytes());
            inputStream.transferTo(outputStream);
            inputStream.close();
            return outputStream.toByteArray();
        });
    }
}

File: test-suite-kotlin-ksp-server-generator/src/main/kotlin/io/micronaut/openapi/test/api/RequestBodyController.kt
package io.micronaut.openapi.test.api

import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Post
import io.micronaut.http.multipart.CompletedFileUpload
import io.micronaut.openapi.test.model.*
import io.micronaut.openapi.test.model.ColorEnum.Companion.fromValue
import reactor.core.publisher.Mono
import java.io.ByteArrayOutputStream

@Controller
open class RequestBodyController : RequestBodyApi {

    @Post(value = "/echo", consumes = ["text/plain"], produces = ["text/plain"])
    fun echo(request: String): String {
        return request
    }

    override fun sendSimpleModel(simpleModel: SimpleModel?): Mono<SimpleModel> {
        return Mono.just(simpleModel!!)
    }

    override fun sendValidatedCollection(requestBody: List<List<String>>?): Mono<Void> {
        return Mono.empty()
    }

    override fun sendListOfSimpleModels(simpleModel: List<SimpleModel>?): Mono<List<SimpleModel>> {
        return Mono.just(simpleModel!!)
    }

    override fun sendModelWithRequiredProperties(modelWithRequiredProperties: ModelWithRequiredProperties?): Mono<ModelWithRequiredProperties> {
        return Mono.just(modelWithRequiredProperties!!)
    }

    override fun sendDateModel(dateModel: DateModel?): Mono<DateModel> {
        return Mono.just(dateModel!!)
    }

    override fun sendEnum(body: String): Mono<ColorEnum> {
        return Mono.just(fromValue(body.replace("\"", "")))
    }

    override fun sendEnumList(colorEnums: List<ColorEnum>): Mono<List<ColorEnum>> {
        return Mono.just(colorEnums)
    }

    override fun sendModelWithMapProperty(modelWithMapProperty: ModelWithMapProperty): Mono<ModelWithMapProperty> {
        return Mono.just(modelWithMapProperty)
    }

    override fun sendModelWithValidatedListProperty(modelWithValidatedListProperty: ModelWithValidatedListProperty): Mono<Void> {
        return Mono.empty()
    }

    override fun sendNestedModel(nestedModel: NestedModel): Mono<NestedModel> {
        return Mono.just(nestedModel)
    }

    override fun sendModelWithInnerEnum(modelWithInnerEnum: ModelWithInnerEnum): Mono<ModelWithInnerEnum> {
        return Mono.just(modelWithInnerEnum)
    }

    override fun sendModelWithDiscriminator(animal: Animal): Mono<Animal> {
        return Mono.just(animal)
    }

    override fun sendBytes(body: ByteArray?): Mono<ByteArray> {
        return Mono.just(body!!)
    }

    override fun sendModelWithEnumList(modelWithEnumList: ModelWithEnumList): Mono<ModelWithEnumList> {
        return Mono.just(modelWithEnumList)
    }

    override fun sendFile(file: CompletedFileUpload?): Mono<ByteArray> {
        return Mono.fromCallable {
            val inputStream = file!!.inputStream
            val outputStream = ByteArrayOutputStream()
            outputStream.write("name: ".toByteArray())
            outputStream.write(file.filename.toByteArray())
            outputStream.write(", content: ".toByteArray())
            inputStream.transferTo(outputStream)
            inputStream.close()
            outputStream.toByteArray()
        }
    }
}
